### PR TITLE
pump/storage: Refactor, encapsulate writing and offset updating in logFile

### PR DESCRIPTION
**Pull request by suzaku**
### What problem does this PR solve? 

When we need to make sure the `writableOffset` is for the file
corresponding to `maxFid`, we may write code like this:

```go
if fid == maxFid {
        // get writableOffset
}
```

To make this threadsafe, we have to introduce a lock.

### What is changed and how it works?

Separate the concern of the struct `valueLog` and `logFile`.

`valueLog` is a higher level interface to the binlog data files, it
manages a directory of `logFile`s.

`logFile` is responsible for saving data to individual files.

By allowing `LogFile` to manage its own `writeOffset`, we can be sure
that the offset is for the "current" file without using a lock.

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

### CODE REVIEW

File: pump/storage/log_test.go [hunk 1]
```diff
@@ -73,6 +74,18 @@ func (lfs *LogFileSuit) TearDownTest(c *check.C) {
...
+func (lfs *LogFileSuit) TestWriteOffset(c *check.C) {
+	dir := c.MkDir()
+	f, err := newLogFile(1024, filepath.Join(dir, "1024.vlog"))
+	c.Assert(err, check.IsNil)
+	c.Assert(f.GetWriteOffset(), check.Equals, int64(0))
+
+	data := make([]byte, 379)
+	err = f.Write(data, true)
+	c.Assert(err, check.IsNil)
+	c.Assert(f.GetWriteOffset(), check.Equals, int64(379))
```
Comment (csuzhangxc) [inline pump/storage/log_test.go]: Is it a _write failed_ case (use `RLIMIT_FSIZE` as before) need?

Comment (suzaku) [reply pump/storage/log_test.go]: That case was already covered elsewhere.

Comment (GregoryIan) [reply pump/storage/log_test.go]: better to migrate the case here?

**Comment by GregoryIan** [comment]
Rest LGTM

File: pump/storage/log_test.go (hunk 1 above)
Comment (suzaku) [reply pump/storage/log_test.go]: Maybe leave that for another PR?

Comment (GregoryIan) [reply pump/storage/log_test.go]: are there something undefined and need to be handled separately in another PR?

**Pull request merged**